# I. Preprocessing
Основную нагрузку несет только pandas, остальные библиотеки можно удалить, либо оставить для последующей визуализации/

In [1]:
# ondoing main libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt
%matplotlib inline

In [2]:
# hiding possible warnings in the system
import warnings
warnings.filterwarnings('ignore')

# II. Add functions
Before start this notebook, please check all comments in function below

In [326]:
#provides a creation of the pressure level dataframe with default delimeter from txt - ";"
#removes the top two extra lines from raw .txt files
#names columns from pattern "names"
def сreatePressureDt(filename):
    names = ["Addr", "Command", "Pressure", "Time", "Date"]
    return pd.read_csv(filename, skiprows = 2, header=None, delimiter=';', names=names)

In [368]:
#provides a creation of the temperature level dataframe with default delimeter from .xlsx - "\t" - "tabulation"
#names columns from pattern "names"
def сreateTempDt(filename):
    names = ["T1, K", "T2, K", "T3, K", "T4, K", "T5, K", "T6, K", "T7, K","T8, K","Date", "Time"]
    dataframe = pd.read_csv(filename, delimiter='\t', names=names)
    return dataframe

In [373]:
#provides a creation of the liquid helium level dataframe with default delimeter from .xlsx - "\t" - "tabulation"
#names columns from pattern "names"
#removes trash "cm" from data cells
def сreateLevelDt(filename):
    names = ["Level, cm", "Date", "Time"]
    dataframe = pd.read_csv(filename, delimiter='\t', names=names)
    dataframe["Level, cm"] = dataframe["Level, cm"].map(lambda x: x.rstrip(" cm"))
    return dataframe

In [329]:
#optional
#concatenate the "Data" and "Time" columns and reformat their into "datetime"
#drops extra "Time" and "Data"
def dateConcat(dataframe):1
    dataframe["Date/Time"] = dataframe['Date'] + " " + dataframe['Time']
    dataframe["Date/Time"] = dataframe["Date/Time"].apply(pd.to_datetime)
    dataframe.drop(['Time', 'Date'], axis=1, inplace=True)

In [330]:
#removes trash words from cells according to pattern "parser_cond"
#removes extra milliseconds from "Time"
def pressureParser(dataframe):
    parser_cond = {"Addr" : "Addr:", "Command" : "Command: ", "Pressure" : "Pressure: ", 
                   "Time" : "Time: ", "Date" : "Date: "}
    for name, parse in parser_cond.items():
        dataframe[name] = dataframe[name].map(lambda x: x.lstrip(parse).rstrip(" kPa"))
        
    dataframe['Time'] = dataframe['Time'].map(lambda x: (x.rstrip(x[-3:])).rstrip("."))

In [331]:
#provides a complete name of column "Pressure" with considering gauge number (1,2 or 4) and unit of measure
def pressureDeterminator(dataframe):
    oldname = "Pressure"
    newname = 'Pressure' + "_" + dataframe["Addr"][0] + ", kPa"
    dataframe.rename(columns={oldname : newname}, inplace=True)
    dataframe[newname] = dataframe[newname].map(lambda x: float(x.replace(',','.')))
    dataframe.drop(['Addr', 'Command'], axis=1, inplace=True)

# III. Pressure data test
Take in your mind that current names of files and their directories can be different on your laptop!

In [332]:
filename_pressure_1 = "Davlenie_start_otkachka_22.30_1.txt"
filename_pressure_2 = "Davlenie_start_otkachka_22.30_2.txt"
filename_pressure_4 = "Davlenie_start_otkachka_22.30_4.txt"

In [333]:
sensor_surf_1 = сreatePressureDt(filename_pressure_1)
sensor_surf_2 = сreatePressureDt(filename_pressure_2)
sensor_surf_4 = сreatePressureDt(filename_pressure_4)

In [334]:
sensor_surf_1

Addr       Command                   Pressure  \
0      Addr: 1    Command: 1    Pressure: 3,4919100 kPa    
1      Addr: 1    Command: 1    Pressure: 3,5085470 kPa    
2      Addr: 1    Command: 1    Pressure: 3,5391730 kPa    
3      Addr: 1    Command: 1    Pressure: 3,5419290 kPa    
4      Addr: 1    Command: 1    Pressure: 3,5612620 kPa    
...         ...           ...                        ...   
34655  Addr: 1    Command: 1    Pressure: 3,6649240 kPa    
34656  Addr: 1    Command: 1    Pressure: 3,6806960 kPa    
34657  Addr: 1    Command: 1    Pressure: 3,6618250 kPa    
34658  Addr: 1    Command: 1    Pressure: 3,6636040 kPa    
34659  Addr: 1    Command: 1    Pressure: 3,6909840 kPa    

                       Time               Date  
0       Time: 22:29:42.033    Date: 27.11.2020  
1       Time: 22:29:43.029    Date: 27.11.2020  
2       Time: 22:29:44.054    Date: 27.11.2020  
3       Time: 22:29:45.066    Date: 27.11.2020  
4       Time: 22:29:46.079    Date: 27.11.2020  
...                     ...                ...  
34655    Time: 8:10:27.936    Date: 28.11.2020  
34656    Time: 8:10:28.950    Date: 28.11.2020  
34657    Time: 8:10:29.949    Date: 28.11.2020  
34658    Time: 8:10:30.956    Date: 28.11.2020  
34659    Time: 8:10:31.962    Date: 28.11.2020  

[34660 rows x 5 columns]

In [335]:
sensor_surf_2

Addr       Command                     Pressure  \
0      Addr: 2    Command: 1      Pressure: 0,1722565 kPa    
1      Addr: 2    Command: 1      Pressure: 0,1740799 kPa    
2      Addr: 2    Command: 1      Pressure: 0,1752014 kPa    
3      Addr: 2    Command: 1      Pressure: 0,1769638 kPa    
4      Addr: 2    Command: 1      Pressure: 0,1794357 kPa    
...         ...           ...                          ...   
34580  Addr: 2    Command: 1    Pressure: -97,1041700 kPa    
34581  Addr: 2    Command: 1    Pressure: -97,1047400 kPa    
34582  Addr: 2    Command: 1    Pressure: -97,1047400 kPa    
34583  Addr: 2    Command: 1    Pressure: -97,1041700 kPa    
34584  Addr: 2    Command: 1    Pressure: -97,1065600 kPa    

                       Time               Date  
0       Time: 22:29:42.345    Date: 27.11.2020  
1       Time: 22:29:43.342    Date: 27.11.2020  
2       Time: 22:29:44.368    Date: 27.11.2020  
3       Time: 22:29:45.378    Date: 27.11.2020  
4       Time: 22:29:46.389    Date: 27.11.2020  
...                     ...                ...  
34580    Time: 8:09:19.008    Date: 28.11.2020  
34581    Time: 8:09:19.991    Date: 28.11.2020  
34582    Time: 8:09:20.958    Date: 28.11.2020  
34583    Time: 8:09:21.953    Date: 28.11.2020  
34584    Time: 8:09:22.957    Date: 28.11.2020  

[34585 rows x 5 columns]

In [336]:
sensor_surf_4

Addr       Command                     Pressure  \
0      Addr: 4    Command: 1    Pressure: -98,3390800 kPa    
1      Addr: 4    Command: 1    Pressure: -98,3396500 kPa    
2      Addr: 4    Command: 1    Pressure: -98,3396500 kPa    
3      Addr: 4    Command: 1    Pressure: -98,3390800 kPa    
4      Addr: 4    Command: 1    Pressure: -98,3385000 kPa    
...         ...           ...                          ...   
34673  Addr: 4    Command: 1    Pressure: -99,5906500 kPa    
34674  Addr: 4    Command: 1    Pressure: -99,5928900 kPa    
34675  Addr: 4    Command: 1    Pressure: -99,5918000 kPa    
34676  Addr: 4    Command: 1    Pressure: -99,5928800 kPa    
34677  Addr: 4    Command: 1    Pressure: -99,5940200 kPa    

                       Time               Date  
0       Time: 22:29:42.655    Date: 27.11.2020  
1       Time: 22:29:43.681    Date: 27.11.2020  
2       Time: 22:29:44.682    Date: 27.11.2020  
3       Time: 22:29:45.687    Date: 27.11.2020  
4       Time: 22:29:46.700    Date: 27.11.2020  
...                     ...                ...  
34673    Time: 8:10:45.647    Date: 28.11.2020  
34674    Time: 8:10:46.634    Date: 28.11.2020  
34675    Time: 8:10:47.646    Date: 28.11.2020  
34676    Time: 8:10:48.652    Date: 28.11.2020  
34677    Time: 8:10:49.652    Date: 28.11.2020  

[34678 rows x 5 columns]

__Как видно, данные содержат излишние обозначения, время имеет излишнюю размерность миллисекунд.__

In [337]:
pressureParser(sensor_surf_1)
pressureParser(sensor_surf_2)
pressureParser(sensor_surf_4)

In [338]:
pressureDeterminator(sensor_surf_1)
pressureDeterminator(sensor_surf_2)
pressureDeterminator(sensor_surf_4)

In [339]:
sensor_surf_1

Pressure_ 1, kPa      Time        Date
0              3.491910  22:29:42  27.11.2020
1              3.508547  22:29:43  27.11.2020
2              3.539173  22:29:44  27.11.2020
3              3.541929  22:29:45  27.11.2020
4              3.561262  22:29:46  27.11.2020
...                 ...       ...         ...
34655          3.664924   8:10:27  28.11.2020
34656          3.680696   8:10:28  28.11.2020
34657          3.661825   8:10:29  28.11.2020
34658          3.663604   8:10:30  28.11.2020
34659          3.690984   8:10:31  28.11.2020

[34660 rows x 3 columns]

In [340]:
sensor_surf_2

Pressure_ 2, kPa      Time        Date
0              0.172257  22:29:42  27.11.2020
1              0.174080  22:29:43  27.11.2020
2              0.175201  22:29:44  27.11.2020
3              0.176964  22:29:45  27.11.2020
4              0.179436  22:29:46  27.11.2020
...                 ...       ...         ...
34580        -97.104170   8:09:19  28.11.2020
34581        -97.104740   8:09:19  28.11.2020
34582        -97.104740   8:09:20  28.11.2020
34583        -97.104170   8:09:21  28.11.2020
34584        -97.106560   8:09:22  28.11.2020

[34585 rows x 3 columns]

In [341]:
sensor_surf_4

Pressure_ 4, kPa      Time        Date
0             -98.33908  22:29:42  27.11.2020
1             -98.33965  22:29:43  27.11.2020
2             -98.33965  22:29:44  27.11.2020
3             -98.33908  22:29:45  27.11.2020
4             -98.33850  22:29:46  27.11.2020
...                 ...       ...         ...
34673         -99.59065   8:10:45  28.11.2020
34674         -99.59289   8:10:46  28.11.2020
34675         -99.59180   8:10:47  28.11.2020
34676         -99.59288   8:10:48  28.11.2020
34677         -99.59402   8:10:49  28.11.2020

[34678 rows x 3 columns]

In [342]:
%%time
dateConcat(sensor_surf_1)

CPU times: user 7.55 s, sys: 1.92 ms, total: 7.55 s
Wall time: 7.55 s


In [343]:
%%time
dateConcat(sensor_surf_2)

CPU times: user 7.6 s, sys: 0 ns, total: 7.6 s
Wall time: 7.59 s


In [344]:
%%time
dateConcat(sensor_surf_4)

CPU times: user 7.62 s, sys: 4.29 ms, total: 7.63 s
Wall time: 7.65 s


__Время на конкатенацию даты и времени не сказать, что очень существенная, но возможно, требуется оптимизация. Требуется уточнение__

__Замечание: у меня стоит ICore 5, 8 Gen, 1,6 ГГц__

In [345]:
sensor_surf_1

Pressure_ 1, kPa           Date/Time
0              3.491910 2020-11-27 22:29:42
1              3.508547 2020-11-27 22:29:43
2              3.539173 2020-11-27 22:29:44
3              3.541929 2020-11-27 22:29:45
4              3.561262 2020-11-27 22:29:46
...                 ...                 ...
34655          3.664924 2020-11-28 08:10:27
34656          3.680696 2020-11-28 08:10:28
34657          3.661825 2020-11-28 08:10:29
34658          3.663604 2020-11-28 08:10:30
34659          3.690984 2020-11-28 08:10:31

[34660 rows x 2 columns]

In [346]:
sensor_surf_2

Pressure_ 2, kPa           Date/Time
0              0.172257 2020-11-27 22:29:42
1              0.174080 2020-11-27 22:29:43
2              0.175201 2020-11-27 22:29:44
3              0.176964 2020-11-27 22:29:45
4              0.179436 2020-11-27 22:29:46
...                 ...                 ...
34580        -97.104170 2020-11-28 08:09:19
34581        -97.104740 2020-11-28 08:09:19
34582        -97.104740 2020-11-28 08:09:20
34583        -97.104170 2020-11-28 08:09:21
34584        -97.106560 2020-11-28 08:09:22

[34585 rows x 2 columns]

In [347]:
sensor_surf_4

Pressure_ 4, kPa           Date/Time
0             -98.33908 2020-11-27 22:29:42
1             -98.33965 2020-11-27 22:29:43
2             -98.33965 2020-11-27 22:29:44
3             -98.33908 2020-11-27 22:29:45
4             -98.33850 2020-11-27 22:29:46
...                 ...                 ...
34673         -99.59065 2020-11-28 08:10:45
34674         -99.59289 2020-11-28 08:10:46
34675         -99.59180 2020-11-28 08:10:47
34676         -99.59288 2020-11-28 08:10:48
34677         -99.59402 2020-11-28 08:10:49

[34678 rows x 2 columns]

# IV. Temperature / level data test
Take in your mind that current names of files and their directories can be different on your laptop!

In [499]:
filename_temp = "TEMP_OTKACHKA_27.11.2020_22.39.xls"
filename_level = "LEVEL_OTKACHKA_27.11.2020_22.39.xls"

In [375]:
temp_surf = сreateTempDt(filename_temp)
level_surf = сreateLevelDt(filename_level)

In [376]:
temp_surf

T1, K  T2, K    T3, K  T4, K  T5, K  T6, K  T7, K  T8, K        Date  \
0      4.476  4.227    4.235  1.203    1.2  1.203  1.203  4.186  27.11.2020   
1      4.477  4.228    4.237  1.203    1.2  1.203  1.203  4.186  27.11.2020   
2      4.477  4.228    3.814  1.203    1.2  1.203  1.203  4.186  27.11.2020   
3      4.477  4.228    3.765  1.203    1.2  1.203  1.203  4.186  27.11.2020   
4      4.477  4.228    3.727  1.203    1.2  1.203  1.203  4.186  27.11.2020   
...      ...    ...      ...    ...    ...    ...    ...    ...         ...   
31304  1.000  1.202  500.000  1.203    1.2  1.203  1.203  1.202  28.11.2020   
31305  1.000  1.202  500.000  1.203    1.2  1.203  1.203  1.202  28.11.2020   
31306  1.000  1.202  500.000  1.203    1.2  1.203  1.203  1.202  28.11.2020   
31307  1.000  1.202  500.000  1.203    1.2  1.203  1.203  1.202  28.11.2020   
31308    NaN    NaN      NaN    NaN    NaN    NaN    NaN    NaN  28.11.2020   

           Time  
0      22:39:17  
1      22:39:19  
2      22:39:20  
3      22:39:21  
4      22:39:22  
...         ...  
31304   8:06:59  
31305   8:07:00  
31306   8:07:01  
31307   8:07:02  
31308   8:07:03  

[31309 rows x 10 columns]

__Как видно, pandas автоматически съедает знаки "+" в ячейках данных__

In [377]:
level_surf

Level, cm        Date      Time
0         108.3  27.11.2020  22:39:39
1         108.3  27.11.2020  22:39:40
2         108.3  27.11.2020  22:39:41
3         108.3  27.11.2020  22:39:42
4         108.3  27.11.2020  22:39:43
...         ...         ...       ...
34161      63.5  28.11.2020   8:09:17
34162      63.5  28.11.2020   8:09:18
34163      63.5  28.11.2020   8:09:19
34164      63.5  28.11.2020   8:09:20
34165      63.5  28.11.2020   8:09:21

[34166 rows x 3 columns]

__Столбец уровней очищается от единиц измерения во время создания. Смотри сreateLevelDt(...)__

In [378]:
%%time
dateConcat(temp_surf)

CPU times: user 7.01 s, sys: 2.7 ms, total: 7.01 s
Wall time: 7.01 s


In [379]:
%%time
dateConcat(level_surf)

CPU times: user 7.62 s, sys: 0 ns, total: 7.62 s
Wall time: 7.64 s


In [380]:
temp_surf

T1, K  T2, K    T3, K  T4, K  T5, K  T6, K  T7, K  T8, K  \
0      4.476  4.227    4.235  1.203    1.2  1.203  1.203  4.186   
1      4.477  4.228    4.237  1.203    1.2  1.203  1.203  4.186   
2      4.477  4.228    3.814  1.203    1.2  1.203  1.203  4.186   
3      4.477  4.228    3.765  1.203    1.2  1.203  1.203  4.186   
4      4.477  4.228    3.727  1.203    1.2  1.203  1.203  4.186   
...      ...    ...      ...    ...    ...    ...    ...    ...   
31304  1.000  1.202  500.000  1.203    1.2  1.203  1.203  1.202   
31305  1.000  1.202  500.000  1.203    1.2  1.203  1.203  1.202   
31306  1.000  1.202  500.000  1.203    1.2  1.203  1.203  1.202   
31307  1.000  1.202  500.000  1.203    1.2  1.203  1.203  1.202   
31308    NaN    NaN      NaN    NaN    NaN    NaN    NaN    NaN   

                Date/Time  
0     2020-11-27 22:39:17  
1     2020-11-27 22:39:19  
2     2020-11-27 22:39:20  
3     2020-11-27 22:39:21  
4     2020-11-27 22:39:22  
...                   ...  
31304 2020-11-28 08:06:59  
31305 2020-11-28 08:07:00  
31306 2020-11-28 08:07:01  
31307 2020-11-28 08:07:02  
31308 2020-11-28 08:07:03  

[31309 rows x 9 columns]

In [381]:
level_surf

Level, cm           Date/Time
0         108.3 2020-11-27 22:39:39
1         108.3 2020-11-27 22:39:40
2         108.3 2020-11-27 22:39:41
3         108.3 2020-11-27 22:39:42
4         108.3 2020-11-27 22:39:43
...         ...                 ...
34161      63.5 2020-11-28 08:09:17
34162      63.5 2020-11-28 08:09:18
34163      63.5 2020-11-28 08:09:19
34164      63.5 2020-11-28 08:09:20
34165      63.5 2020-11-28 08:09:21

[34166 rows x 2 columns]

# V. Concatanation dataframes

### Way 1 : Merge way
Попробуем соединить базы данных по SQL стилю, через понятие объединение множеств "how="outer"", сравнивать будем по столбцу "Date/Time". По идее должны собраться все данные, пустые автоматически заполнятся NaNами.

In [496]:
union_pres_df = pd.merge(sensor_surf_1, sensor_surf_2, on="Date/Time", how="outer")
union_pres_df = pd.merge(union_pres_df, sensor_surf_4, on="Date/Time", how="outer")

Отметим, что метод pd.merge() может производить слияние одновременно только с двумя базами данных, поэтому осуществляется слияние вначале двух, а потом слияние с третьей БД

In [497]:
union_pres_df

Pressure_ 1, kPa           Date/Time  Pressure_ 2, kPa  \
0              3.491910 2020-11-27 22:29:42          0.172257   
1              3.508547 2020-11-27 22:29:43          0.174080   
2              3.539173 2020-11-27 22:29:44          0.175201   
3              3.541929 2020-11-27 22:29:45          0.176964   
4              3.561262 2020-11-27 22:29:46          0.179436   
...                 ...                 ...               ...   
35128               NaN 2020-11-28 08:10:45               NaN   
35129               NaN 2020-11-28 08:10:46               NaN   
35130               NaN 2020-11-28 08:10:47               NaN   
35131               NaN 2020-11-28 08:10:48               NaN   
35132               NaN 2020-11-28 08:10:49               NaN   

       Pressure_ 4, kPa  
0             -98.33908  
1             -98.33965  
2             -98.33965  
3             -98.33908  
4             -98.33850  
...                 ...  
35128         -99.59065  
35129         -99.59289  
35130         -99.59180  
35131         -99.59288  
35132         -99.59402  

[35133 rows x 4 columns]

In [484]:
# Get the column names
cols = union_pres_df.columns.tolist()
cols

['Pressure_ 1, kPa', 'Date/Time', 'Pressure_ 2, kPa', 'Pressure_ 4, kPa']

In [485]:
# Reorder the columns
cols = cols[:1]+cols[2:]+cols[1:2]
cols

['Pressure_ 1, kPa', 'Pressure_ 2, kPa', 'Pressure_ 4, kPa', 'Date/Time']

In [486]:
union_pres_df = union_df[cols]

In [487]:
union_pres_df

Pressure_ 1, kPa  Pressure_ 2, kPa  Pressure_ 4, kPa  \
0              3.491910          0.172257         -98.33908   
1              3.508547          0.174080         -98.33965   
2              3.539173          0.175201         -98.33965   
3              3.541929          0.176964         -98.33908   
4              3.561262          0.179436         -98.33850   
...                 ...               ...               ...   
35128               NaN               NaN         -99.59065   
35129               NaN               NaN         -99.59289   
35130               NaN               NaN         -99.59180   
35131               NaN               NaN         -99.59288   
35132               NaN               NaN         -99.59402   

                Date/Time  
0     2020-11-27 22:29:42  
1     2020-11-27 22:29:43  
2     2020-11-27 22:29:44  
3     2020-11-27 22:29:45  
4     2020-11-27 22:29:46  
...                   ...  
35128 2020-11-28 08:10:45  
35129 2020-11-28 08:10:46  
35130 2020-11-28 08:10:47  
35131 2020-11-28 08:10:48  
35132 2020-11-28 08:10:49  

[35133 rows x 4 columns]

In [488]:
union_pres_temp_df = pd.merge(union_pres_df, temp_surf, on="Date/Time", how="outer")
union_pres_temp_df

Pressure_ 1, kPa  Pressure_ 2, kPa  Pressure_ 4, kPa  \
0              3.491910          0.172257         -98.33908   
1              3.508547          0.174080         -98.33965   
2              3.539173          0.175201         -98.33965   
3              3.541929          0.176964         -98.33908   
4              3.561262          0.179436         -98.33850   
...                 ...               ...               ...   
35128               NaN               NaN         -99.59065   
35129               NaN               NaN         -99.59289   
35130               NaN               NaN         -99.59180   
35131               NaN               NaN         -99.59288   
35132               NaN               NaN         -99.59402   

                Date/Time  T1, K  T2, K  T3, K  T4, K  T5, K  T6, K  T7, K  \
0     2020-11-27 22:29:42    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1     2020-11-27 22:29:43    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2     2020-11-27 22:29:44    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3     2020-11-27 22:29:45    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4     2020-11-27 22:29:46    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
...                   ...    ...    ...    ...    ...    ...    ...    ...   
35128 2020-11-28 08:10:45    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
35129 2020-11-28 08:10:46    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
35130 2020-11-28 08:10:47    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
35131 2020-11-28 08:10:48    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
35132 2020-11-28 08:10:49    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

       T8, K  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN  
...      ...  
35128    NaN  
35129    NaN  
35130    NaN  
35131    NaN  
35132    NaN  

[35133 rows x 12 columns]

In [489]:
union = pd.merge(union_pres_temp_df, level_surf, on="Date/Time", how="outer")
union

Pressure_ 1, kPa  Pressure_ 2, kPa  Pressure_ 4, kPa  \
0              3.491910          0.172257         -98.33908   
1              3.508547          0.174080         -98.33965   
2              3.539173          0.175201         -98.33965   
3              3.541929          0.176964         -98.33908   
4              3.561262          0.179436         -98.33850   
...                 ...               ...               ...   
35129               NaN               NaN         -99.59065   
35130               NaN               NaN         -99.59289   
35131               NaN               NaN         -99.59180   
35132               NaN               NaN         -99.59288   
35133               NaN               NaN         -99.59402   

                Date/Time  T1, K  T2, K  T3, K  T4, K  T5, K  T6, K  T7, K  \
0     2020-11-27 22:29:42    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1     2020-11-27 22:29:43    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2     2020-11-27 22:29:44    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3     2020-11-27 22:29:45    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4     2020-11-27 22:29:46    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
...                   ...    ...    ...    ...    ...    ...    ...    ...   
35129 2020-11-28 08:10:45    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
35130 2020-11-28 08:10:46    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
35131 2020-11-28 08:10:47    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
35132 2020-11-28 08:10:48    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
35133 2020-11-28 08:10:49    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

       T8, K Level, cm  
0        NaN       NaN  
1        NaN       NaN  
2        NaN       NaN  
3        NaN       NaN  
4        NaN       NaN  
...      ...       ...  
35129    NaN       NaN  
35130    NaN       NaN  
35131    NaN       NaN  
35132    NaN       NaN  
35133    NaN       NaN  

[35134 rows x 13 columns]

In [490]:
# Get the column names
cols_union = union.columns.tolist()
cols_union

['Pressure_ 1, kPa',
 'Pressure_ 2, kPa',
 'Pressure_ 4, kPa',
 'Date/Time',
 'T1, K',
 'T2, K',
 'T3, K',
 'T4, K',
 'T5, K',
 'T6, K',
 'T7, K',
 'T8, K',
 'Level, cm']

In [491]:
# Reorder the columns
cols_union = cols_union[:3]+cols_union[4:]+cols_union[3:4]
# union = union[cols_union]
cols_union

['Pressure_ 1, kPa',
 'Pressure_ 2, kPa',
 'Pressure_ 4, kPa',
 'T1, K',
 'T2, K',
 'T3, K',
 'T4, K',
 'T5, K',
 'T6, K',
 'T7, K',
 'T8, K',
 'Level, cm',
 'Date/Time']

In [493]:
union = union[cols_union]
union

Pressure_ 1, kPa  Pressure_ 2, kPa  Pressure_ 4, kPa  T1, K  T2, K  \
0              3.491910          0.172257         -98.33908    NaN    NaN   
1              3.508547          0.174080         -98.33965    NaN    NaN   
2              3.539173          0.175201         -98.33965    NaN    NaN   
3              3.541929          0.176964         -98.33908    NaN    NaN   
4              3.561262          0.179436         -98.33850    NaN    NaN   
...                 ...               ...               ...    ...    ...   
35129               NaN               NaN         -99.59065    NaN    NaN   
35130               NaN               NaN         -99.59289    NaN    NaN   
35131               NaN               NaN         -99.59180    NaN    NaN   
35132               NaN               NaN         -99.59288    NaN    NaN   
35133               NaN               NaN         -99.59402    NaN    NaN   

       T3, K  T4, K  T5, K  T6, K  T7, K  T8, K Level, cm           Date/Time  
0        NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-27 22:29:42  
1        NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-27 22:29:43  
2        NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-27 22:29:44  
3        NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-27 22:29:45  
4        NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-27 22:29:46  
...      ...    ...    ...    ...    ...    ...       ...                 ...  
35129    NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-28 08:10:45  
35130    NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-28 08:10:46  
35131    NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-28 08:10:47  
35132    NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-28 08:10:48  
35133    NaN    NaN    NaN    NaN    NaN    NaN       NaN 2020-11-28 08:10:49  

[35134 rows x 13 columns]

__В итоге, как и ожидалось, получили БД с уникальными Date/Time полями, пустые поля заполнены NaNами__

# VI. Data output -> Excel file
Coming soon...__This topic can write the dataframes into generated excel file in exist sheet. 
If .xlsx file does not exist, module automatically creates it with specified name in pd.ExcelWriter(...) and put it in current directory with this notebook__

In [498]:
# Модуль может открывать файл из концретной директории, а также открывать файл на запись или дозапись
# По сути, работа, как с обычным файлом в Питоне или Си

# Файл на запись - по умолчанию: 
# pd.ExcelWriter('OutputData.xls')
# Файл на дозапись: 
# pd.ExcelWriter('OutputData.xls', mode="A")

In [495]:
%%time
with pd.ExcelWriter('OutputData.xls') as writer:
#     sensor_surf_1.to_excel(writer,index=False, startcol=0)
#     sensor_surf_2.to_excel(writer,index=False, startcol=2)
#     sensor_surf_4.to_excel(writer,index=False, startcol=4)
#     union_df.to_excel(writer,index=False)
#     level_surf.to_excel(writer,index=False, startcol=4)
#     temp_surf.to_excel(writer,index=False, startcol=6)
    union.to_excel(writer,index=False)
    writer.save()

CPU times: user 6.5 s, sys: 39.7 ms, total: 6.54 s
Wall time: 6.56 s


__Время "перекачки" 35134 строк и 13 колонок заняло всего 6,56 секунд. Однако, у меня стоит SSD, на HDD скорости будут менее веселыми, стоит это иметь в виду__

__Требуется проверка на новых форматах файлов, после надстроек Маши в установке. Скорее всего придется фиксить существующие функции__